harvester/cloud-provider-harvester
# feat: get node name from guest agent

https://github.com/harvester/harvester/issues/4947

### Test steps

1. Create a harvester with https://github.com/harvester/harvester/pull/5019.
2. Create VMImage and VM Network.
3. Import the harvester to a Rancher cluster.
4. Use harvester kubeconfig to run [deploy/generate_addon.sh](https://github.com/harvester/cloud-provider-harvester/blob/master/deploy/generate_addon.sh). The `cluster_name` is guest cluster name which we will create later. The `namespace` is guest cluster VM namespace.
```
./deploy/generate_addon <cluster_name> <namespace>
```
5. Set a RKE2 with `External` cloud provider.
<img width="404" alt="Screenshot 2024-01-24 at 2 32 53 PM" src="https://github.com/harvester/cloud-provider-harvester/assets/4927361/face6a21-3e13-40b2-bb8c-810cc4eac87f">

6. Copy `write_files` from the step 4 to User Data. Also, setting a `hostname`.
<img width="934" alt="Screenshot 2024-01-24 at 2 40 47 PM" src="https://github.com/harvester/cloud-provider-harvester/assets/4927361/fb33d98a-2448-405c-ad84-02a89a3f8c03">

7. Create the RKE2 cluster and wait for message `waiting for cluster agent to connect`.
<img width="630" alt="Screenshot 2024-01-24 at 2 33 15 PM" src="https://github.com/harvester/cloud-provider-harvester/assets/4927361/73b0a4f5-87c7-49f9-ba43-b0d8b556013b">

8. Use `ssh` to login the VM and install `helm`.
9. In the VM, set kubeconfig.
```
sudo chmod 666 /etc/rancher/rke2/rke2.yaml
export KUBECONFIG=/etc/rancher/rke2/rke2.yaml
```
10. In the VM, clone the harvester charts and install harvester-cloud-provider.
```
git clone https://github.com/harvester/charts
cd charts/charts/harvester-cloud-provider/
helm install harvester-cloud-provider . --namespace kube-system \
  --set image.repository=frankyang/harvester-cloud-provider \
  --set image.tag=447fef20-amd64 \
  --set global.cattle.clusterName=<cluster_name>
```
11. Check the RKE2 cluster can be ready.
